# Understanding Variables and Data Types

**Contents**
1. Variables
2. Naming rules and conventions
3. Comments
4. Common data types
5. The `print()` function
6. `type()` and `isinstance()`

---
# 1. Variables

Variables are labeled boxes used for storing and referencing data of different types, such as numbers or text.

To create a variable, write its name on the left, then the assignment operator (`=`), then the value you want to store on the right.

In [1]:
name = 'John Doe'
age = 25

print(name)
print(age)

John Doe
25


Here `name` is the variable that holds the value `'John Doe'`.

That value is a **string**. Strings are a series of characters used to represent text, written in single or double quotes.

---
# 2. Naming Rules and Conventions

## Rules
Breaking any of these raises a `SyntaxError`.

- Names can **only start** with a letter or an underscore (`_`), **not** a number
- Names can only contain alphanumeric characters (`a-z`, `A-Z`, `0-9`) and underscores (`_`)
- Names can **not** be one of Python's reserved keywords, such as `if`, `class`, and `def`

Names are also **case-sensitive**: `age`, `Age`, and `AGE` are three different variables.

In [2]:
age = 25
Age = 30
AGE = 35

print(age, Age, AGE)   # three separate variables

25 30 35


In [3]:
_valid_name = 'starts with an underscore'
valid_name_2 = 'numbers are fine after the first character'
print(_valid_name)
print(valid_name_2)

# Uncomment any line below to see the SyntaxError:
# 2nd_name = 'starts with a number'
# my-name = 'contains a hyphen'
# class = 'reserved keyword'

starts with an underscore
numbers are fine after the first character


## Snake case
Variable names should be lowercase, with words separated by underscores. This is called **snake case**.

In [4]:
my_variable_name = 'freeCodeCamp'
print(my_variable_name)

freeCodeCamp


## Be descriptive
Choose names that say what the value is. To store a user's age, `user_age` is better than `age` or an abbreviation.

Avoid single-letter names such as `x`, because they communicate no purpose or meaning.

In [5]:
x = 25           # unclear: 25 what?
user_age = 25    # clear

print(user_age)

25


---
# 3. Comments

A comment starts with the pound symbol (`#`). Python ignores everything after the `#` on that line.

In [6]:
# This is a single-line comment
print('Comments are ignored')  # a comment can also follow code on the same line

Comments are ignored


---
# 4. Common Data Types

## Integer
A whole number without decimals, for example `10` or `-5`.

In [7]:
my_integer_var = 10
print('Integer:', my_integer_var)

Integer: 10


## Float
A number with a decimal point, like `4.41` or `-0.4`.

In [8]:
my_float_var = 4.50
print('Float:', my_float_var)

Float: 4.5


## String
A sequence of characters enclosed in single or double quotation marks, like `'Hello world!'`.

In [9]:
my_string_var = 'hello'
print('String:', my_string_var)

String: hello


## Boolean
A true or false type, written as `True` or `False` (capitalized).

In [10]:
my_boolean_var = True
print('Boolean:', my_boolean_var)

Boolean: True


---
# 5. The `print()` Function

`print()` is a **built-in function**, which means you can use it without defining it yourself. It sends text and other values to the terminal.

The value inside the parentheses is called an **argument**. Here the string `'Hello world!'` is the argument passed to `print()`.

In [11]:
print('Hello world!')

Hello world!


## Multiple arguments
`print()` can show several values separated by commas. Python automatically puts a space between them.

In [12]:
name = 'John Doe'
age = 25

print('Name:', name, 'Age:', age)

Name: John Doe Age: 25


---
# 6. `type()` and `isinstance()`

## type()
Shows the class, or type, of a variable.

In [13]:
developer = 'Devin'
print(type(developer))

<class 'str'>


`<class 'str'>` means `developer` is a string.

Here are all four data types so far:

In [14]:
print(type(my_integer_var))
print(type(my_float_var))
print(type(my_string_var))
print(type(my_boolean_var))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


## isinstance()
Checks whether a variable matches a specific data type. It takes a value and the type to check against, then returns a Boolean.

This is useful when you need to verify a variable's type before performing operations on it.

In [15]:
print(isinstance(my_integer_var, int))
print(isinstance(my_integer_var, str))
print(isinstance(my_string_var, str))

True
False
True


## Checking multiple types at once
Pass the types inside parentheses (a tuple). It returns `True` if the value matches **any** of them.

In [16]:
print(isinstance(my_float_var, (int, float)))
print(isinstance(my_string_var, (int, float)))

True
False


---
# Quick Reference

| Type | Name in Python | Example |
|---|---|---|
| Integer | `int` | `10` |
| Float | `float` | `4.5` |
| String | `str` | `'hello'` |
| Boolean | `bool` | `True` |

| Syntax | What it does |
|---|---|
| `user_age = 25` | Create a variable |
| `# note` | Comment (ignored by Python) |
| `print(a, b)` | Show values, separated by a space |
| `type(user_age)` | Show the variable's type |
| `isinstance(user_age, int)` | `True` if the variable is that type |
| `isinstance(user_age, (int, float))` | `True` if it is any of the listed types |